src/orange/surf/detail/FerrariSolver.hh

```diff
@@ -0,0 +1,347 @@
+        Real5 abcde,
+        SurfaceState on_surface,
+        real_type tolerance = Tolerance<real_type>::sqrt_quadratic());
+
+  public:
+    // Construct w/ given tolerance. Uses quadratic tolerance by default.
+    inline CELER_FUNCTION
+    FerrariSolver(real_type tolerance = Tolerance<real_type>::sqrt_quadratic());
+
+    // Solver for fully general case
+    inline CELER_FUNCTION Intersections operator()(Real5 abcde) const;
+
+  private:
+    //// DATA ////
+    // Soft zero for biquadratic and degenerate cubic detection
+    static inline SoftZero<real_type> const soft_zero_;
+
+    //// UTIL ////
+    // Try to place real at given index in list, return next free index
+    static inline CELER_FUNCTION int
+    place_root(Intersections& roots, real_type new_root, int free_index);
+
+    // Find roots of special reduced quartic which is biquadratic
+    static inline CELER_FUNCTION Intersections
+    calc_biquadratic_roots(real_type qb, real_type p, real_type r);
+
+    // Find dominant root of normalized cubic
+    static inline CELER_FUNCTION real_type
+    dominant_root_normalized_cubic(real_type b, real_type c, real_type d);
+
+    // Find real quadratic roots
+    static inline CELER_FUNCTION Real2
+    real_roots_normalized_quadratic(real_type b, real_type c);
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Find all positive roots for quartic surfaces using Ferrari-Cardano method.
+ *
+ * This method allows the user to manually specify if the particle starts on
+ * the surface, and use 0 instead of e for accuracy.
+ */
+CELER_FUNCTION auto FerrariSolver::solve_general(Real5 abcde,
+                                                 SurfaceState on_surface,
+                                                 real_type tolerance)
+    -> Intersections
+{
+    FerrariSolver solve(tolerance);
+    auto selected_abcde = abcde;
+    if (on_surface == SurfaceState::on)
+    {
+        auto [a, b, c, d, e] = abcde;
+        selected_abcde = {a, b, c, d, 0};
+    }
+    return solve(selected_abcde);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a solver instance with a specified tolerance for degenerate cases,
+ * such as the particle starting on the surface.
+ */
+CELER_FUNCTION
+FerrariSolver::FerrariSolver(real_type tolerance)
+{
+    SoftZero soft_zero_{tolerance};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find all positive roots of the polynomial with given a, b, c, d, e.
+ *
+ * The polynomial takes the form:
+ * \f[
+   ax^4 + bx^3 + cx^2 + dx + e = 0.
+ *\f]
+ * Where the given array abcde corresponds to {a, b, c, d, e}.
+ * Replaces negative or complex roots with sentry value no_intersection().
+ */
+CELER_FUNCTION auto FerrariSolver::operator()(Real5 abcde) const
```

Comment:
Since you immediately unpack these (and the coefficients aren't modified like the "general" case), you could/should use a const reference here. Since it's all inline, the compiler *should* optimize it to the same thing, though.